API provides a way to for user to pay for a subscription

AlexSaxena opened the pull request:
[PT-link](https://www.pivotaltracker.com/story/show/169253777)

## PR content
In this PR we are adding a feature where a User can pay for a subscription and be upgraded to the Subscriber role

## This PR contains:
- Stripe gem
-  Stripe-ruby-mock gem
- Payments Controller
- Unit tests
- Upgrade User role to Subscriber role

## What have we learned?
- How to set up Stripe
- How to utilize the Stripe-ruby-mock gem
- Improved understanding of RSpec and related technologies

AlexSaxena committed:
Adds requests_spec for Stripe transaction.

Commit:
Generate Payment controller & add stripe gems & update request spec

AlexSaxena committed:
Adds Customer and Charge in payment Controller.

AlexSaxena committed:
Renames payment controller to be plural,
Configures routes for payment.

Commit:
Add sad path to request spec

Add Stripe credentials.
Add render message to Payments controller for failed transactions.

AlexSaxena committed:
Removes leftover comments and binding.pry

Commit:
Rename file to match user story

faraznaeem commented on app/controllers/payments_controller.rb:
```diff
@@ -0,0 +1,30 @@
+  def create
+    if params[:stripeToken]
+
+      amount = 1000
```
why is the value here hard coded?

shsteindorsson replied:
The plan was to save any subscription related things for the user-story where we add the subscription options.

faraznaeem commented on app/controllers/payments_controller.rb:
```diff
@@ -0,0 +1,30 @@
+      charge = Stripe::Charge.create(
+        customer: customer.id,
+        amount: amount,
+        description: 'New Fake News Subscriber',
```
perhaps use the user's name here instead?

shsteindorsson replied:
When authenticating with `user.name` instead of `user.id`, we receive a Stripe error message:
```
Stripe::InvalidRequestError:
       Must provide source or customer.
```

faraznaeem commented on config/routes.rb:
```diff
@@ -1,6 +1,7 @@
 Rails.application.routes.draw do
   namespace :v1, defaults: { format: :json } do 
     resources :articles, only: [:index, :create, :show, :update]
+    resources :payments, only: [:index, :create]
```
where is the `:index` action?

AlexSaxena replied:
Resolved

faraznaeem commented on spec/requests/v1/api_provides_a_way_to_add_subscription_spec.rb:
```diff
@@ -0,0 +1,32 @@
+  before { StripeMock.start }
+  after { StripeMock.stop }
+  
+  describe 'User would like to pay for subscritption' do
```
typo subscription

AlexSaxena replied:
Resolved

faraznaeem commented on spec/requests/v1/api_provides_a_way_to_add_subscription_spec.rb:
```diff
@@ -0,0 +1,32 @@
+                                      }, headers: headers
+      user.reload
+      expect(response.status).to eq 200
+      expect(user.role).to eq('subscriber')
```
is it good to have multiple `expectations` in the same `it` block?

AlexSaxena replied:
Resolved

AlexSaxena committed:
Refactors code in Spec_file for multiple it-blocks
Refactors routes,
Corrects Typo.

Commit:
Accept incoming changes

Commit:
Add begin-rescue statements to payments controller

Add improved error handling with begin-rescue.
Add expectations for messages to request spec.

AlexSaxena committed:
Removes leftover comments

AlexSaxena committed:
Updates spec test with error status and message.
